File: a3_finance/a3_finance/report/payment_advice_to_bank/payment_advice_to_bank.py
# Copyright (c) 2025, Acube
# License: see license.txt

import frappe

def execute(filters=None):
    if not filters:
        filters = {}

    # Convert month & year to int
    filters["payroll_month"] = int(filters.get("payroll_month") or 0)
    filters["payroll_year"] = int(filters.get("payroll_year") or 0)

    # Employment type filter
    emp_type_filter = ""
    if filters.get("employment_type"):
        if filters["employment_type"] == "Regular Employees":
            emp_type_filter = "AND e.employment_type IN ('Officers', 'Workers')"
        else:
            emp_type_filter = "AND e.employment_type = %(employment_type)s"

    columns = [
        {"label": "Sl No", "fieldname": "sl_no", "fieldtype": "Data", "width": 120},
        {"label": "Employee ID", "fieldname": "employee", "fieldtype": "Data", "width": 120},
        {"label": "Employee Name", "fieldname": "employee_name", "fieldtype": "Data", "width": 180},
        {"label": "Bank Account", "fieldname": "bank_account", "fieldtype": "Data", "width": 180},
        {"label": "Net Pay", "fieldname": "net_pay", "fieldtype": "Currency", "width": 120},
       
    ]

    data = frappe.db.sql("""
        SELECT  
            e.employee_number AS employee,
            e.employee_name,
                    
            e.bank_ac_no AS bank_account,
            ss.net_pay
        FROM `tabSalary Slip` ss
        JOIN `tabEmployee` e ON ss.employee = e.name
        WHERE 
            MONTH(ss.start_date) = %(payroll_month)s
            AND YEAR(ss.start_date) = %(payroll_year)s
            {emp_type_filter}
    """.format(emp_type_filter=emp_type_filter), filters, as_dict=True)

    # Total footer
        # Add Sl No
    for i, row in enumerate(data, start=1):
        row["sl_no"] = i
    total = sum(row["net_pay"] for row in data)
    if data:
        data.append({
            "employee": "",
            "employee_name": "TOTAL",
            "bank_account": "",
            "net_pay": total
        })

    return columns, data

File: a3_finance/a3_finance/report/payment_advice_to_bank/payment_advice_to_bank.js
// © 2025 ACUBE

frappe.query_reports["Payment Advice To Bank"] = {

    filters: [
        {
            fieldname: "company",
            label: "Company",
            fieldtype: "Link",
            options: "Company"
        },
        {
            fieldname: "payroll_month",
            label: "Payroll Month",
            fieldtype: "Select",
            options: [
                { label: "January", value: 1 }, { label: "February", value: 2 }, { label: "March", value: 3 },
                { label: "April", value: 4 },   { label: "May", value: 5 },      { label: "June", value: 6 },
                { label: "July", value: 7 },    { label: "August", value: 8 },   { label: "September", value: 9 },
                { label: "October", value: 10 },{ label: "November", value: 11 },{ label: "December", value: 12 }
            ],
            default: new Date().getMonth() + 1
        },
        {
            fieldname: "payroll_year",
            label: "Payroll Year",
            fieldtype: "Data",
            default: new Date().getFullYear().toString()
        },
        {
            fieldname: "employment_type",
            label: "Employment Type",
            fieldtype: "Link",
            options: "Employment Type"
        },
        {
            fieldname: "prepared_by",
            label: "Prepared by",
            fieldtype: "Link",
            options: "Employee"
        },
        // {
        //     fieldname: "checked_by",
        //     label: "Checked by",
        //     fieldtype: "Link",
        //     options: "Employee"
        // },
        // {
        //     fieldname: "verified_by",
        //     label: "Verified by",
        //     fieldtype: "Link",
        //     options: "Employee"
        // },
        {
            fieldname: "approved_by",
            label: "Approved by",
            fieldtype: "Link",
            options: "Employee"
        }
    ],

    onload(report) {
        report.page.add_inner_button("Print Report", async function () {

            const data = frappe.query_report.data || [];
            if (!data.length) {
                frappe.msgprint("No data to print.");
                return;
            }

            const html = await frappe.query_reports["Payment Advice To Bank"].printable_html(data);

            let win = window.open("", "_blank");
            win.document.write(`<html><head><title>Payment Advice</title></head><body>${html}</body></html>`);
            win.document.close();
            win.print();
        });
    },

    async printable_html(data) {

        // Collect filter values
        const f = name => frappe.query_report.get_filter_value(name);
        const filters = {
            prepared: f("prepared_by"),
            checked:  f("checked_by"),
            verified: f("verified_by"),
            approved: f("approved_by"),
            company:  f("company")
        };

        // -- Fetch employee details ------------------------------------------------
        const emp_ids = [filters.prepared, filters.checked, filters.verified, filters.approved].filter(Boolean);
        let emp_map = {};

        if (emp_ids.length) {
            const employees = await frappe.db.get_list("Employee", {
                filters: { name: ["in", emp_ids] },
                fields: ["name", "employee_name", "designation"]
            });

            employees.forEach(e => {
                emp_map[e.name] = `${e.employee_name}${e.designation ? " - " + e.designation : ""}`;
            });
        }

        const prepared = emp_map[filters.prepared] || "";
        const checked  = emp_map[filters.checked]  || "";
        const verified = emp_map[filters.verified] || "";
        const approved = emp_map[filters.approved] || "";

        // -- Date --------------------------------------------
        const month = f("payroll_month");
        const year  = f("payroll_year");

        const monthName = [
            "", "January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"
        ][month];

        // -- HTML OUTPUT -------------------------------------------------------------
        return `
            <div style="text-align:center; font-family:Arial;">
                <h2>${filters.company || ""}</h2>
                <div style="font-size:13px;">CHACKAI, BEACH P.O., AIRPORT ROAD, THIRUVANANTHAPURAM</div>
                <h3 style="margin-top:10px;">Payment Advice To Bank</h3>
                <h4>${monthName} ${year}</h4>
            </div>

            <table border="1" cellspacing="0" cellpadding="6" width="100%"
                style="font-family:Arial; font-size:12px; border-collapse:collapse; margin-top:20px;">
                <thead style="background:#f5f5f5; font-weight:bold;">
                    <tr>
                        <th>Sl No</th>
                        <th>Employee ID</th>
                        <th>Employee Name</th>
                        <th>Bank Account</th>
                        <th style="text-align:right;">Net Pay</th>
                    </tr>
                </thead>
                <tbody>
                    ${data.map(r => `
                        <tr>
                            <td>${r.sl_no || ""}</td>
                            <td>${r.employee || ""}</td>
                            <td>${r.employee_name || ""}</td>
                            <td>${r.bank_account || ""}</td>
                            <td style="text-align:right;">${r.net_pay || 0}</td>
                        </tr>
                    `).join("")}
                </tbody>
            </table>

            <br><br>

          <table width="100%" style="font-family: Arial; font-size: 12px; text-align: center;">
    <tr>
        <!-- Left: Authorized -->
        <td width="50%">
            <b>Authorized Signatory & Designation</b>
            <div style="height:55px;"></div>
        </td>

      
    </tr>

    <tr>
        <!-- Prepared below Authorized -->
        <td>
            
            <div style="height:55px;"></div>
            ${prepared}
        </td>

        <td width="50%">
           
            <div style="height:55px;"></div>
            ${approved}
        </td>

        <!-- Empty cell to keep structure -->
        <td></td>
    </tr>
</table>

        `;
    }
};
